### PR TITLE
Serialize resource attributes in target info for PrometheusHttpServer

### LINKED ISSUE #4552: Prometheus exporter resource attributes
The [spec says](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/data-model.md#resource-attributes-1) the prometheus exporter SHOULD convert resource attributes to the target info metric family.

We don't currently do this, but we should.

### PR DESCRIPTION
**Pull request by jack-berg**
Resolves #4552.

### CODE REVIEW

File: exporters/prometheus/src/integrationTest/java/io/opentelemetry/exporter/prometheus/PrometheusCollectorIntegrationTest.java [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+import org.slf4j.LoggerFactory;
+import org.testcontainers.containers.BindMode;
+import org.testcontainers.containers.GenericContainer;
+import org.testcontainers.containers.output.Slf4jLogConsumer;
+import org.testcontainers.containers.wait.strategy.Wait;
+import org.testcontainers.junit.jupiter.Testcontainers;
+import org.testcontainers.utility.DockerImageName;
+
+@Testcontainers(disabledWithoutDocker = true)
+class PrometheusCollectorIntegrationTest {
```
Comment (jack-berg) [inline exporters/prometheus/src/integrationTest/java/io/opentelemetry/exporter/prometheus/PrometheusCollectorIntegrationTest.java:53]: I added this integration test to codify the behavior of the round trip: `java sdk -> java prometheus exporter -> collector prometheus receiver -> collector otlp exporter` 

The collector behavior still has some rough edges, so I expect the behavior of this test to evolve as the prometheus exporter works towards stability.

File: exporters/prometheus/src/integrationTest/java/io/opentelemetry/exporter/prometheus/PrometheusCollectorIntegrationTest.java [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+    // TODO: when otel-collector-contrib:0.55.0 is released, update to include resource attributes
+    // which it will scrape from target info and include on the OTLP resource
+    assertThat(resourceMetrics.getResource().getAttributesList())
+        .containsExactlyInAnyOrder(
+            stringKeyValue(ResourceAttributes.SERVICE_NAME.getKey(), "app"),
+            stringKeyValue(ResourceAttributes.HOST_NAME.getKey(), "host.testcontainers.internal"),
+            stringKeyValue("job", "app"),
+            stringKeyValue("instance", "host.testcontainers.internal:" + prometheusPort),
+            stringKeyValue("port", String.valueOf(prometheusPort)),
+            stringKeyValue("scheme", "http"));
```
Comment (jack-berg) [inline exporters/prometheus/src/integrationTest/java/io/opentelemetry/exporter/prometheus/PrometheusCollectorIntegrationTest.java:128]: Here we see the resource attributes the collector prometheus receiver adds:
- It sets `service.name` from the name of the target
- Notably, it doesn't yet set `service.instance.id`, `net.host.name`, `net.host.port`, or `http.scheme` as described in the [spec](https://github.com/open-telemetry/opentelemetry-specification/blame/main/specification/metrics/data-model.md#L1291-L1300). 
- As noted in the TODO, the target info labels will appear as resource attributes in the `0.55.0` of the collector. Here's the [PR](https://github.com/open-telemetry/opentelemetry-specification/blame/main/specification/metrics/data-model.md#L1291-L1300) that adds that behavior. I tested it locally for now.

Comment (jack-berg) [reply exporters/prometheus/src/integrationTest/java/io/opentelemetry/exporter/prometheus/PrometheusCollectorIntegrationTest.java:128]: Left a [comment](https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/9967#issuecomment-1170532799) on a relevant issue in the collector seeking to understand why the behavior is different from the spec.

File: exporters/prometheus/src/main/java/io/opentelemetry/exporter/prometheus/Serializer.java [hunk 3]
```diff
@@ -115,8 +117,12 @@ final void write(Collection<MetricData> metrics, OutputStream output) throws IOE
                             PrometheusType.forMetric(metric)),
                     LinkedHashMap::new,
                     Collectors.toList()));
+    Optional<Resource> optResource = metrics.stream().findFirst().map(MetricData::getResource);
```
Comment (anuraaga) [inline exporters/prometheus/src/main/java/io/opentelemetry/exporter/prometheus/Serializer.java:120]: Just check isempty and set to null when it is possibly with ternary. Don't need a stream to do that ;)

**Review by anuraaga — COMMENTED** [review]
I don't think we should make any code change until getting more clarity on opt-in configuration here. Or at least then start by making it all opt-in.

Spec says SHOULD for this conversion so we are compliant even if not doing it until we have a way to not blow things up by recording too much resource

**Comment by jack-berg** [comment]
@anuraaga I'm not sure I'm following - is there something in the spec that is ambiguous about _which_ resource attributes are included? This [section](https://github.com/open-telemetry/opentelemetry-specification/blame/main/specification/metrics/data-model.md#L1400) feels fairly straight forward. 

Additionally, this behavior is purely additive right now. Users / scrapers that don't use the target info can just ignore it so not sure of the downside.

**Comment by anuraaga** [comment]
The spec does not take into consideration the size and cardinality of a typical resource so I think it needs clarification. This is similar to not populating resource in Zipkin, I would expect a general definition of what part of the resource to use for cardinality-sensitive backends.

Even if a user doesn't need target_info, Prometheus will still scrape and save it, possibly causing load problems, won't it?

**Comment by jack-berg** [comment]
> Even if a user doesn't need target_info, Prometheus will still scrape and save it

Yes that's true.

But prometheus should already be ascribing unique instance ids to the metrics of each scrape target, which should have a one-to-one relationship with unique sets of resource attributes. I believe that means that including target_info shouldn't increase cardinality beyond what already exists.

**Comment by jack-berg** [comment]
Talked about this in spec issue [2640](https://github.com/open-telemetry/opentelemetry-specification/issues/2640#issuecomment-1190496761) and I've come away thinking no change is needed to the specification and we should move forward with this PR. The TL;DR is that prometheus query syntax allows you to join the target_info attributes at query time, so this data _is_ valuable. The cardinality concerns we discussed are probably small, and users can use `metric_relabel_rules` to drop high cardinality attributes if needed. No need to reinvent the wheel with additional sdk prometheus exporter configuration IMO.

**Comment by anuraaga** [comment]
If every user is basically required to use the relabel rules to drop info, it seems to be a guaranteed UX burden. We do know of certain attributes like instance ID and pid that are more or less required to drop - it seems like a new target_info time series on every server restart would not be efficient even if it can be joined (not sure, can it really join across series with label changes?).

As we see, OTel resource does not have a notion of ensuring low cardinality, unlike metric attributes, and in the future high cardinality items may be added at any time. So a deny list is not really safe, and I guess with a relabel config, every user needs to use `labelkeep` to define an accept list of labels not including the problematic ones. If every single user has to do this (my hypothesis is this is true), then I guess the exporter could just take care of it to make things easier, especially since we're the most familiar with the semantic conventions.

**Comment by jack-berg** [comment]
> not sure, can it really join across series with label changes?

I believe yes, but users will likely be joining on `instance`, so a lot will depend on how they identify the instances they scrape. 

The point I disagree on is the likelihood that _most_ users will be impacted by the increased cardinality of this. The max number of timeseries produced by target_info, regardless of how many high cardinality resource attributes are present is 1 per SDK instance. Prometheus can handle [a relatively large number](https://prometheus.io/docs/prometheus/1.8/storage/#settings-for-high-numbers-of-time-series) of timeseries, and estimates support for ~200k with default modest heap allocation of 2gb. Assuming the user isn't already bumping into limits, let's say they have 10% headroom available, or 20k timeseries to spare. With the default retention of 15 days, they could be running 10 apps, with 10 instances each, and each restarting 13 times a day before they exceed 20k series for target_info. And a 2gb heap prometheus instance is pretty modest resource allocation for an environment with 100 instances (10 apps * 10 instances / app).

So while I can't say with high certainty, my SWAG is that idk, something like less than 2% of users will notice the cardinality of target_info. Seems much more likely they'll first notice cardinality issues from their instruments. And IMO, the potential value they could get with joins to target_info outweighs the reduced UX a small population of users will experience.

I don't feel especially strongly about this, and am happy to let the prometheus wg sort out a decision. Ultimately I think the OTLP exporter is far superior to the prometheus exporter and will always be an advocate for that 😛

**Comment by anuraaga** [comment]
I think I've added all I can so leave you to make the decision :) One final suggestion is it would probably be good to copy what the target_info would look like with the javaagent to Prometheus wg to get their approval, if they approve it should be fine. Note that if it's spring boot the process.command_line will be orders of magnitude shorter than other typical commands that pass a classpath.

**Comment by gouthamve** [comment]
Adding to this, the _main_ thing driving Prometheus memory usage is _active series_, which is typically series which recieved a sample in the last 2-3hrs. If a series didn't get a sample, it is removed from memory. This means with a 20K series headroom, they need to see ~20K restarts in _3hrs_ to hit the limit.

I think the Prometheus WG is in consensus to add all resource attributes to `target_info`. If required, have a denylist of attributes.

> copy what the target_info would look like with the javaagent.

I am going to look into this as I am not super familiar with it. Added it to the agenda for the next WG meeting, but will try to resolve it before then.

**Pull request closed without merging**